storage/storage.rs

```diff
@@ -20,243 +21,111 @@ use std::{
-    }
-
-    for (existing_id, existing_keyspace_index) in keyspaces_index.iter().enumerate() {
-        if let Some(existing_index) = existing_keyspace_index {
-            let keyspace = &keyspaces[existing_index.0 as usize];
-            if keyspace.name() == name {
-                return Err(NameExists { name });
-            }
-            if existing_id == keyspace_id.id().0 as usize {
-                return Err(IdExists { new_name: name, id: keyspace_id.id().0, existing_name: keyspace.name() });
-            }
-        }
-    }
-    Ok(())
-}
-
-#[derive(Debug)]
-pub enum KeyspaceValidationError {
-    IdReserved { name: &'static str, id: u8 },
-    IdTooLarge { name: &'static str, id: u8, max_id: u8 },
-    NameExists { name: &'static str },
-    IdExists { new_name: &'static str, id: u8, existing_name: &'static str },
-}
-
-impl fmt::Display for KeyspaceValidationError {
-    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        match self {
-            Self::NameExists { name, .. } => write!(f, "keyspace '{name}' is defined multiple times."),
-            Self::IdReserved { name, id, .. } => write!(f, "reserved keyspace id '{id}' cannot be used for new keyspace '{name}'."),
-            Self::IdTooLarge { name, id, max_id, .. } => write!(
-                f, "keyspace id '{id}' cannot be used for new keyspace '{name}' since it is larger than maximum keyspace id '{max_id}'.",
-            ),
-            Self::IdExists { new_name, id, existing_name, .. } => write!(
-                f,
-                "keyspace id '{}' cannot be used for new keyspace '{}' since it is already used by keyspace '{}'.",
-                id, new_name, existing_name
-            ),
-        }
-    }
-}
-
-impl Error for KeyspaceValidationError {}
-
-fn db_options() -> Options {
-    let mut options = Options::default();
-    options.create_if_missing(true);
-    options.create_missing_column_families(true);
-    options.enable_statistics();
-    // TODO optimise per-keyspace
-    options
-}
-
-fn recover_keyspaces<KS: KeyspaceSet>(
-    storage_dir: impl AsRef<Path>,
-) -> Result<(Vec<Keyspace>, [Option<KeyspaceId>; KEYSPACE_MAXIMUM_COUNT]), StorageRecoverError> {
-    use StorageRecoverError::{KeyspaceOpen, KeyspaceValidation};
-
-    let path = storage_dir.as_ref();
-    let mut keyspaces = Vec::new();
-    let mut keyspaces_index = core::array::from_fn(|_| None);
-    let options = db_options();
-    for keyspace_id in KS::iter() {
-        validate_new_keyspace(keyspace_id, &keyspaces, &keyspaces_index)
-            .map_err(|error| KeyspaceValidation { source: error })?;
-        keyspaces.push(Keyspace::open(path, keyspace_id, &options).map_err(|error| KeyspaceOpen { source: error })?);
-        keyspaces_index[keyspace_id.id().0 as usize] = Some(KeyspaceId(keyspaces.len() as u8 - 1));
-    }
-    Ok((keyspaces, keyspaces_index))
-}
-
-impl<D> MVCCStorage<D> {
+impl<Durability> MVCCStorage<Durability> {
```

Comment:
Let's create a `Checkpoint` module.
We can move both create and load methods into this new module.

`Chekpoint.load(path)` makes a stateful `Checkpoint` struct, which you can query for number of applied commits, etc.

`Storage.open()` can then load the last checkpoint, and if any commits were applied, immediately create a new one before returning the completed storage.

I think we can also try to hide away the usage of isolation manager into a temporary isolation manager used only for the duration of recovery.